File: tree/binaryTree.go
package main

import(
	// "bufio"
	"fmt"
	// "os"
	"strconv"
	"github.com/k0kubun/pp"
	"../procon"
)

type Node struct{
	parent, left, right int
}

func main() {
	n, err := strconv.Atoi(procon.StrStdin())
	if err != nil {
		panic(err)
	}
	binaryTree := make([]Node, n, n)
	for i := 0; i < n; i++ {
		binaryTree[i].parent = -1
	}
	for i := 0; i < n; i++ {
		nodeInfo, err := procon.SplitIntStdin(" ")
		if err != nil {
			panic(err)
		}
		if nodeInfo[1] != -1 {
			binaryTree[nodeInfo[0]].left = nodeInfo[1]
			binaryTree[nodeInfo[1]].parent = nodeInfo[0]
		}
		if nodeInfo[2] != -1 {
			binaryTree[nodeInfo[0]].right = nodeInfo[2]
			binaryTree[nodeInfo[2]].parent = nodeInfo[0]
		}
	}
	pp.Println(binaryTree)
	fmt.Println(height(binaryTree))
}

func height(binaryTree []Node) []int {
	heights := make([]int, len(binaryTree), len(binaryTree))
	for i:= 0; i < len(binaryTree); i++ {
		heights[i] = nodeHeight(binaryTree[i], binaryTree)
	}
	return heights
}

func nodeHeight(node Node, binaryTree []Node) int {
	if node.parent == -1 { return 1}
	height := 1 + nodeHeight(binaryTree[node.parent], binaryTree)
	return height
}

File: tree/binaryTree.cpp
#include<cstdio>
#define MAX 10000
#define NIL -1

struct Node {int parent, left, right}

Node T[MAX];
int n, D[MAX], H[MAX];

void setDepth(int um, int d) {
  if (u== NIL) return;
  D[u] = d;
  setDepth(T[u].left, d + 1);
  setDepth(T[u].right, d + 1);

}

int setHeight(int u) {
  int h1 = 0, h2 = 0;
  if (T[u].left != NIL)
    h1 = SetHeight(T[u].left) + 1;
  if (T[u].right != NIL)
    h2 = setHeight(T[u].right) + 1;
  return H[u] = (h1 > h2 ? h1 : h2);
}

int getSibling(int u) {
  if (T[u].parent == NIL) return NIL;
  if (T[u].parent.left != u && T[T[u].parent].left != NIL )
    return T[T[u].parent].left;
  if (T[T[u].parent].right != u && T[T[u].parent].right != NIL)
    return T[T[u].parent].right;
  return NIL;
}
